sandbox/cgroup/tracking.go

```diff
@@ -0,0 +1,243 @@
+				isSessionBus = false
+				conn, err = dbusutil.SystemBus()
+				if err != nil {
+					logger.Debugf("system bus is not available: %s", err)
+					return ErrCannotTrackProcess
+				}
+				logger.Debugf("using system bus now, session bus could not activate systemd --user")
+				goto tryAgain
+			}
+			return ErrCannotTrackProcess
+		}
+		return err
+	}
+	// We may have created a transient scope but due to a kernel design,
+	// in specific situation when we are in a cgroup owned by one user,
+	// and we want to run a process as a different user, *and* systemd is
+	// older than 238, then this can silently fail.
+	//
+	// Verify the effective tracking cgroup and check that our scope name
+	// is contained therein.
+	path, err := cgroupProcessPathInTrackingCgroup(pid)
+	if err != nil {
+		return err
+	}
+	if !strings.Contains(path, unitName) {
+		logger.Debugf("systemd could not associate process %d with transient scope %s", pid, unitName)
+		return ErrCannotTrackProcess
+	}
+	return nil
+}
+
+// ConfirmSystemdServiceTracking checks if systemd tracks this process as a snap service.
+//
+// Systemd is placing started services, both user and system, into appropriate
+// tracking groups. Given a security tag we can confirm if the current process
+// belongs to such tracking group and thus could be identified by snapd as
+// belonging to a particular snap and application.
+//
+// If the application process is not tracked then ErrCannotTrackProcess is returned.
+func ConfirmSystemdServiceTracking(securityTag string) error {
+	pid := osGetpid()
+	path, err := cgroupProcessPathInTrackingCgroup(pid)
+	if err != nil {
+		return err
+	}
+	unitName := fmt.Sprintf("%s.service", securityTag)
+	if !strings.Contains(path, unitName) {
+		return ErrCannotTrackProcess
+	}
+	return nil
+}
+
+func sessionOrMaybeSystemBus(uid int) (isSessionBus bool, conn *dbus.Conn, err error) {
+	// The scope is created with a DBus call to systemd running either on
+	// system or session bus. We have a preference for session bus, as this is
+	// where applications normally go to. When a session bus is not available
+	// and the invoking user is root, we use the system bus instead.
+	//
+	// It is worth noting that hooks will not normally have a session bus to
+	// connect to, as they are invoked as descendants of snapd, and snapd is a
+	// service running outside of any session.
+	conn, err = dbusutil.SessionBus()
+	if err == nil {
+		logger.Debugf("using session bus")
+		return true, conn, nil
+	}
+	logger.Debugf("session bus is not available: %s", err)
+	if uid == 0 {
+		logger.Debugf("falling back to system bus")
+		conn, err = dbusutil.SystemBus()
+		if err != nil {
+			logger.Debugf("system bus is not available: %s", err)
+		} else {
+			logger.Debugf("using system bus now, session bus was not available")
+		}
+	}
+	return isSessionBus, conn, err
+}
+
+var errDBusUnknownMethod = errors.New("org.freedesktop.DBus.Error.UnknownMethod")
```

Comment:
using the dbus error name as the message is a bit strange, we usually try to provide some description:

 * unknown dbus object method
 * dbus name has owner
 * ...

Reply:
I see, it's because they are never returned outside the package, I don't have  a strong opinion but if left as is, they probably need a comment explaining that.  another approach would be to a error struct like

type handledDBusError struct {
  dbusError string
}

Reply:
actually we do surface these errors in the logs at line:

https://github.com/snapcore/snapd/pull/8829/files#diff-4706d268ffa8aa3999d09cac05b3ced2R80

so they should have some real message.